feat: add i18n error message support for TextField

vursen opened the pull request:
## Description

The PR introduces TextFieldI18n with methods for setting error messages and updates the TextField's validation logic to show these error messages when validation fails.

> [!NOTE]
> It's still possible to use the `setErrorMessage` method to configure a single static error message. This error message will be displayed for all constraints and will take priority over any i18n error messages that are also set. However, when more than one error message is needed, i18n should be used or manual validation mode should be enabled.

Depends on
- https://github.com/vaadin/flow-components/pull/6322

Part of #4618 

## Type of change

- [x] Feature

vursen commented on vaadin-text-field-flow-parent/vaadin-text-field-flow/src/main/java/com/vaadin/flow/component/textfield/TextField.java:
```diff
@@ -337,4 +360,169 @@ protected void onAttach(AttachEvent attachEvent) {
+         *
+         * @return the error message or {@code null} if not set
+         * @see TextField#isRequiredIndicatorVisible()
+         * @see TextField#setRequiredIndicatorVisible(boolean)
```
**question:** Should we reference `setRequired(boolean)` and `isRequired()` here as well?

vursen commented on vaadin-text-field-flow-parent/vaadin-text-field-flow/src/main/java/com/vaadin/flow/component/textfield/TextField.java:
```diff
@@ -304,31 +288,71 @@ public String getValue() {
+
+        ValidationResult maxLengthResult = ValidationUtil
+                .validateMaxLengthConstraint(getMaxLengthErrorMessage(), value,
+                        hasMaxLength() ? getMaxLength() : null);
```
**note:** `getMaxLength()` returns `0` if the maxLength property is `null`. I added this check to ensure the max length is only considered when the user has explicitly set it with `setMinLength(int minLength)`.

tomivirkki commented on vaadin-flow-components-shared-parent/vaadin-flow-components-base/src/main/java/com/vaadin/flow/component/shared/ValidationUtil.java:
```diff
@@ -170,4 +170,72 @@ public static <V extends Comparable<V>> ValidationResult validateMinConstraint(
+     * @param maxLength
+     *            the maximum allowed length
+     * @return {@code ValidationResult.ok()} if the value is longer than or
+     *         equal to the minimum length, {@code ValidationResult.error()}
```
`minimum` -> `maximum`

vursen replied:
Thanks, fixed this and also another typo.

tomivirkki commented on vaadin-text-field-flow-parent/vaadin-text-field-flow/src/main/java/com/vaadin/flow/component/textfield/TextField.java:
```diff
@@ -304,31 +288,71 @@ public String getValue() {
+            }
+        }
+
+        ValidationResult maxLengthResult = ValidationUtil
```
I wonder if it would make sense to reuse `validateMaxConstraint`?
```java
ValidationResult maxLengthResult = ValidationUtil.validateMaxConstraint(
                getMaxLengthErrorMessage(), value.length(),
                hasMaxLength() ? getMaxLength() : null);
```

tomivirkki replied:
I don't have a strong preference for either approach so this isn't a blocker for the PR

DiegoCardoso commented on vaadin-text-field-flow-parent/vaadin-text-field-flow/src/main/java/com/vaadin/flow/component/textfield/TextField.java:
```diff
@@ -337,4 +361,167 @@ protected void onAttach(AttachEvent attachEvent) {
+                "The i18n properties object should not be null");
+    }
+
+    private String getRequiredErrorMessage() {
+        return Optional.ofNullable(i18n)
+                .map(TextFieldI18n::getRequiredErrorMessage).orElse("");
+    }
+
+    private String getMinLengthErrorMessage() {
+        return Optional.ofNullable(i18n)
+                .map(TextFieldI18n::getMinLengthErrorMessage).orElse("");
+    }
+
+    private String getMaxLengthErrorMessage() {
+        return Optional.ofNullable(i18n)
+                .map(TextFieldI18n::getMaxLengthErrorMessage).orElse("");
+    }
+
+    private String getPatternErrorMessage() {
+        return Optional.ofNullable(i18n)
+                .map(TextFieldI18n::getPatternErrorMessage).orElse("");
+    }
```
Could use the same approach as suggested in https://github.com/vaadin/flow-components/pull/6365#discussion_r1675358980

vursen replied:
Updated.

vursen commented:
During our internal discussions, we found a way to keep error messages that are set with `setErrorMessage`. They now take priority over i18n error messages, which allows developers to use the old straightforward approach to set a single error message for all constraints. I've updated the PR and also removed the breaking mark.

sissbruecker commented on vaadin-text-field-flow-parent/vaadin-text-field-flow/src/main/java/com/vaadin/flow/component/textfield/TextField.java:
```diff
@@ -40,12 +47,13 @@
     private boolean manualValidationEnabled = false;
 
+    private String customErrorMessage;
+    private String constraintErrorMessage;
```
What do you think about extracting this somehow? There could be a class that:
- Manages custom + constraint error message states
- Updates the effective error message property on the component when either is set

It wouldn't do a whole lot, but it looks like you have to replicate the logic quite a few times for different components. There's also `checkValidity` which is always the same though that one needs to access more stuff from the component. Anyway, maybe something to think about, we can also do this after updating all components and then see if there is something that works for all of them.

vursen replied:
Are you suggesting extracting this only for text fields or for all field components?

vursen replied:
I guess we could have something like `ErrorMessageController`:

```java
class ErrorMessageController {
  private Component component;

  public ErrorMessageController(HasValidation component) {
    this.component = component;
  }

  public void setCustomErrorMessage() {
    ...
    this.updateErrorMessageProperty();    
  }

  public void setConstraintErrorMessage() {
    ...
    this.updateErrorMessageProperty();
  }

  public void updateErrorMessageProperty() {
    this.component.getElement().setProperty("errorMessage");
  }
}
```

vursen replied:
> There's also checkValidity which is always the same though that one needs to access more stuff from the component. 

I discussed this with @yuriy-fix some time ago, and we concluded that it would currently complicate things especially since the whole i18n needs to be passed to that helper class or the class would need to provide an API to abstract itself from i18n. We agreed to revisit this in the future when we get to implementing error message providers.

sissbruecker replied:
Yep, something like what you suggested with the controller. It seems like that would be applicable to all components, but not sure. But since there are other potential opportunities for code reuse, let's maybe finish changing the components first, and then see what we can extract as common logic?

vursen replied:
Agreed 👍

vursen commented on vaadin-text-field-flow-parent/vaadin-text-field-flow/src/main/java/com/vaadin/flow/component/textfield/TextField.java:
```diff
@@ -304,31 +328,79 @@ public String getValue() {
-     * constraints using browser development tools.
+     * Validates the current value against the constraints and sets the
+     * {@code invalid} property and the {@code errorMessage} property using the
+     * error messages defined in the i18n object.
```
Done.
